### PR TITLE
Combine peaks

### PR DESCRIPTION
**Pull request by jorainer**
This PR adds a `combinePeaks,Spectra` method that (as the name says) combines peaks **within each** spectrum. The *old* `combinePeaks` function (used by `combineSpectra`) was renamed to `combinePeaksData` as described in #292.

### CODE REVIEW

**Comment by jorainer** [comment]
> Looking [here](https://github.com/rformassspectrometry/Spectra/pull/296/files#diff-8294df09cfc7e83db3834388455ffe70e18e97d27e5b55c5e5e83774ccd95903R507), the message says that the backend will be changed to one supporting data modification. Is this done implicitly by the `.combine_spectra()`?

yes, it's done [here](https://github.com/rformassspectrometry/Spectra/pull/296/files#diff-8294df09cfc7e83db3834388455ffe70e18e97d27e5b55c5e5e83774ccd95903L446)

File: R/Spectra.R [hunk 1]
```diff
@@ -499,11 +500,28 @@ NULL
...
+#'   in their m/z values that is smaller than the maximal acceptable difference
+#'   defined by `ppm` and `tolerance`. Parameters `intensityFun` and `mzFun`
+#'   allow to define functions to aggregate the intensity and m/z values for
+#'   each such group of peaks. With `weighted = TRUE` (the default), the m/z
+#'   value of the combined peak is calculated using an intensity-weighted mean
+#'   and parameter `mzFun` is ignored. The [MsCoreUtils::group()] function is
+#'   used for the grouping of mass peaks. Parameter `msLevel.` allows to
+#'   define selected MS levels for which peaks should be combined. This
+#'   function returns a `Spectra` with the same number of spectra than the
+#'   input object, but with eventually combined peaks within each spectrum.
```
[the remark this answers is not in the document]
Comment (jorainer) [reply R/Spectra.R:513]: OK, will change that.

File: R/Spectra.R [hunk 2]
```diff
@@ -499,11 +500,28 @@ NULL
...
+#'   If additional peak variables (other than `"mz"` and `"intensity"`) are
+#'   present, the values for the first peak (i.e. the peak with the lowest
```
[the remark this answers is not in the document]
Comment (jorainer) [reply R/Spectra.R]: I changed that now in [this](https://github.com/rformassspectrometry/Spectra/pull/296/commits/7780b67741e3a4a06fbaf9b5d30f936b369a181a) commit: for merged peaks, peak variable values are replaced with `NA`.

File: R/Spectra.R [hunk 3]
```diff
@@ -499,11 +500,27 @@ NULL
...
+#'   allow to define functions to aggregate the intensity and m/z values for
+#'   each such group of peaks. With `weighted = TRUE` (the default), the m/z
+#'   value of the combined peak is calculated using an intensity-weighted mean
+#'   and parameter `mzFun` is ignored. The [MsCoreUtils::group()] function is
+#'   used for the grouping of mass peaks. Parameter `msLevel.` allows to
+#'   define selected MS levels for which peaks should be combined. This
+#'   function returns a `Spectra` with the same number of spectra than the
+#'   input object, but with possibly combined peaks within each spectrum.
+#'   Additional peak variables (other than `"mz"` and `"intensity"`) are
+#'   dropped (i.e. their values are replaced with `NA`) for combined peaks. See
```
Comment (lgatto) [inline R/Spectra.R:515]: Should this not be:

Additional peak variables (other than `"mz"` and `"intensity"`) are dropped (i.e. their values are replaced with `NA`) for combined peaks unless they are constant across the combined peaks.

File: R/peaks-functions.R [hunk 4]
```diff
@@ -562,3 +562,50 @@ joinPeaksGnps <- function(x, y, xPrecursorMz = NA_real_,
...
+            mzs, ints, FUN = function(m, i) weighted.mean(m, i + 1),
+            SIMPLIFY = FALSE, USE.NAMES = FALSE), use.names = FALSE)
+    else mzs <- vapply1d(mzs, mzFun)
+    ints <- vapply1d(ints, intensityFun)
+    if (ncol(x) > 2) {
+        idx <- match(seq_len(grps[lg]), grps)
+        lst <- lapply(x[!colnames(x) %in% c("mz", "intensity")], function(z) {
+            z <- z[idx]
+            z[mrg] <- NA
+            z
```
Comment (lgatto) [inline R/peaks-functions.R:605]: My comment above might not be correct... you always set the additional peaks variables to 0 upon combining, even if they are constant? Is this meant to stay like that, or do you consider updating in the future. Imagine for example the situation where you want to combine groups of peaks based on one of these additional peaks variable - dropping would be a shame.

**Review by lgatto — COMMENTED** [review]
Hopefully final discussion about how extra peaks variables are combined. Otherwise, all good for me.

**Comment by lgatto** [comment]
Here's the code that deals with invariant columns in QFeatures https://github.com/rformassspectrometry/QFeatures/blob/master/R/reduce.R

File: R/Spectra.R (hunk 3 above)
Comment (jorainer) [reply R/Spectra.R:515]: The current implementation just drops them without checking if they are constant (for peaks that are combined).

**Review by lgatto — APPROVED** [review]

File: R/Spectra.R (hunk 3 above)
Comment (jorainer) [reply R/Spectra.R]: changed it [now](https://github.com/rformassspectrometry/Spectra/pull/296/commits/74ffe66f78578a35313e10cc034eeb0d095a4123) to match that: values for peak variables are only replaced with `NA` if, across the combined peaks, they are not constant.

**Pull request merged**